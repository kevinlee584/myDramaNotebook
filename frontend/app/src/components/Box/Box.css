.box {
    width: 225px;
    height: 225px;
    border: 1px solid black; 
    margin: auto;
    margin-top: 15px;
    margin-bottom: 10px;
    box-shadow: rgba(6, 24, 44, 0.4) 0px 0px 0px 2px, rgba(6, 24, 44, 0.65) 0px 4px 6px -1px, rgba(255, 255, 255, 0.08) 0px 1px 0px inset;
    border-radius: 10px;
    overflow: hidden;
    background-color: black;
    position: relative;
}

.video-block {
    height: 100%;
    width: 100%;
    text-decoration: none;
    position: absolute;
    display: block;
}

.video-image {
    height: 80%;
    width: 100%;
    border-radius: 10px 10px 0 0;
    position: absolute;
    left: 0%;
    top: 0%;
}

.video-name-blocker {
    display: flex;
    width: auto;
    height: 20%;
    position: absolute;
    overflow: hidden;
    bottom: 0%;
}

.video-name {
    background-color: #2e2f31;
    height: 100%;
    margin: 0;
    align-items: center;
    overflow: hidden;
    width: auto;
    display: flex;
    border-radius: 0 0 10px 10px;
    position: relative;
}

.video-name p {
    color: aliceblue;
    float: left;
    min-width: 225px;
    text-align: left;
    white-space: nowrap;
    position: relative;
    margin-left: 10px;
    
}

.video-name:hover {
    animation: marquee 5s linear infinite;
    animation-delay: 0.5s;
}

.save-button {
    background-color: #2e2f31;
    height: 20%;
    width: 20%;
    position: absolute;
    right: 0%;
    bottom: 0%;
    display: flex;
    align-items: center;
    justify-content: center;
    border: 0px;
}

.save-button svg {
    height: 70% !important;
    width: 70% !important;
}

@keyframes marquee {
    0% { left: 0; }
    100% { left: calc(-50%); }
}