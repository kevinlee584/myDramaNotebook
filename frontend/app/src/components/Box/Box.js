import React, { useState } from "react";
import RecordService from "../../service/RecordService"
import Save from "./save.svg"
import Saved from "./saved.svg"
import "./Box.css";

const Box = ({ drama, record }) => {

    const { providerName, imageUrl, videoUrl, name } = drama
    const [isSaved, setIsSaved] = useState(record.find(e => e.providerName == providerName && e.name == name) != undefined)

    const saveFun = (event) => {
        event.preventDefault()
        if (isSaved) {
            RecordService.removeDrama(providerName, name)
            .then(_res => {
                setIsSaved(false)
                let index = record.findIndex(e => e.providerName == providerName && e.name == name)
                record.splice(index, 1)
                })
        }else {
            RecordService.addDrama(providerName, name)
            .then(_res => {
                setIsSaved(true)
                record.push(drama)
            })
        }
    }

    return (
    <div className="box">
        <a href={videoUrl} className="video-block" target="_blank">
            <>
                <img src={imageUrl} className="video-image"></img>
                <div className="video-name-blocker">
                    <div className="video-name">
                        <p>{name}</p>
                        <p>{name}</p>
                    </div>
                </div>
                <div className="save-botton" onClick={saveFun}>
                    { isSaved ? <Saved></Saved> : <Save></Save> }
                </div>
            </>
        </a>
    </div>
)










    // const { providerName, imageUrl, videoUrl, name } = drama
    // const isSave = record.find(e => e.providerName == providerName && e.name == name) != undefined
    // const [saved, setSaved] = useState(isSave)

    // const clicked = (event) => {
    //     event.preventDefault()
    //     if (saved) {
    //         fetch(`${config.backend.url}/user/remove`, 
    //             {method: "DELETE", headers: headers, 
    //             body: JSON.stringify({provider: providerName, name: name})})
    //         .then(res => {
    //             setSaved(false)
    //             let index = record.findIndex(e => e.providerName == providerName && e.name == name)
    //             record.splice(index, 1)
    //             setRecord(record)
    //         })
    //     }
    //     else {
    //         fetch(`${config.backend.url}/user/save`, 
    //             {method: "POST", 
    //             headers: headers, 
    //             body: JSON.stringify({provider: providerName, name: name})})
    //         .then(res => {
    //             setSaved(true)
    //             record.push(drama)
    //             setRecord(record)
    //         })
    //     }
    // }

    // return (
    //     <div className="box">
    //         <a href={videoUrl} className="video-block" target="_blank">
    //             <>
    //                 <img src={imageUrl} className="video-image"></img>
    //                 <div className="video-name-blocker">
    //                     <div className="video-name">
    //                         <p>{name}</p>
    //                         <p>{name}</p>
    //                     </div>
    //                 </div>
    //                 <div className="save-botton" onClick={clicked}>
    //                     { saved ? <Saved></Saved> : <Save></Save> }
    //                 </div>
    //             </>
    //         </a>
    //     </div>
    // )
};

export default Box;