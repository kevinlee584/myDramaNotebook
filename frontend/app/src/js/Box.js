import React, { useEffect, useState } from "react";
import Save from "../static/icons/save.svg"
import Saved from "../static/icons/saved.svg"
import config from '../../../config'
import "../css/box.css";

let headers = {
    "Content-Type": "application/json",
    "Accept": "application/json, text/plain;charset=UTF-8"
}

const Box = ({ drama, record, setRecord }) => {

    const { providerName, imageUrl, videoUrl, name } = drama
    const isSave = record.find(e => e.providerName == providerName && e.name == name) != undefined
    const [saved, setSaved] = useState(isSave)

    const clicked = (event) => {
        event.preventDefault()
        if (saved) {
            fetch(`${config.backend.url}/user/remove`, 
                {method: "DELETE", headers: headers, 
                body: JSON.stringify({provider: providerName, name: name})})
            .then(res => {
                setSaved(false)
                let index = record.findIndex(e => e.providerName == providerName && e.name == name)
                record.splice(index, 1)
                setRecord(record)
            })
        }
        else {
            fetch(`${config.backend.url}/user/save`, 
                {method: "POST", 
                headers: headers, 
                body: JSON.stringify({provider: providerName, name: name})})
            .then(res => {
                setSaved(true)
                record.push(drama)
                setRecord(record)
            })
        }
    }

    return (
        <div className="box">
            <a href={videoUrl} className="video-block" target="_blank">
                <>
                    <img src={imageUrl} className="video-image"></img>
                    <div className="video-name-blocker">
                        <div className="video-name">
                            <p>{name}</p>
                            <p>{name}</p>
                        </div>
                    </div>
                    <div className="save-botton" onClick={clicked}>
                        { saved ? <Saved></Saved> : <Save></Save> }
                    </div>
                </>
            </a>
        </div>
    )
};

export default Box;