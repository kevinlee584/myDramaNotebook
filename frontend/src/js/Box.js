import React, { useEffect } from "react";

import "../css/box.css";
  
const Box = ({ imageUrl, videoUrl, videoName }) => {

    return (
        <div className="box">
            <a href={videoUrl} className="video-block" target="_blank">
                <>
                    <img src={imageUrl} className="video-image"></img>
                    <div className="video-name-blocker">
                        <div className="video-name">
                            <p>{videoName}</p>
                            <p>{videoName}</p>
                        </div>
                    </div>
                </>
            </a>
        </div>
    )
};

export default Box;