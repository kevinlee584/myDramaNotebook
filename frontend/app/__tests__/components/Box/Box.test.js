import {render, fireEvent, act} from '@testing-library/react'
import {describe, test, expect, jest} from '@jest/globals';

jest.mock('../../../src/service/RecordService', () => { return {
    addDrama: function (_providerName, _dramaName) { return Promise.resolve("add"); },
    removeDrama: function (_providerName, _dramaName) { return Promise.resolve("remove") }
}})
jest.mock('../../../src/components/Box/save.svg', () => () => (<div>Save</div>))
jest.mock('../../../src/components/Box/saved.svg', () => () => (<div>Saved</div>))
import Box from '../../../src/components/Box/Box'


describe(Box, () => {
    const drama =  { 
        providerName: "test",
        name: "test" ,
        imageUrl: "testUrl", 
        videoUrl: "testUrl", 
    }

    test("renders correctly", () => {
        const box = render(<Box drama={drama} record={[]}></Box>)
        expect(box.asFragment()).toMatchSnapshot()
    })

    test("should be Save state", () => {
        const box = render(<Box drama={drama} record={[]}></Box>)
        expect(box.queryByRole('button', {name: "Save"})).not.toBeNull()
    })

    test("should be Saved state", () => {
        const box = render(<Box drama={drama} record={[drama]}></Box>)
        expect(box.queryByRole('button', {name: "Saved"})).not.toBeNull()
    })

    test("should change Save to Saved state", async () => {
        const box = render(<Box drama={drama} record={[]}></Box>)
        const btn = box.queryByRole('button', {name: "Save"})

        await act( async () => {
            fireEvent.click(btn)
        });
        expect(box.queryByRole('button', {name: "Saved"})).not.toBeNull()
    })


})