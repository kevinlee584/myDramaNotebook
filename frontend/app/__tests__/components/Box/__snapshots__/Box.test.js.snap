// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders correctly 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <a
      class="video-block"
      href="testUrl"
      target="_blank"
    >
      <img
        class="video-image"
        src="testUrl"
      />
      <div
        class="video-name-blocker"
      >
        <div
          class="video-name"
        >
          <p>
            test
          </p>
          <p>
            test
          </p>
        </div>
      </div>
      <button
        class="save-button"
      >
        <div>
          Save
        </div>
      </button>
    </a>
  </div>
</DocumentFragment>
`;
